src/mozanalysis/experiment.py

```diff
@@ -0,0 +1,545 @@
+            `metric_list`. Then one column per sanity-check metric.
+            Columns:
+                - client_id (str, optional): Not necessary for
+                    "happy path" analyses.
+                - branch (str): The client's branch
+                - [metric 1]: The client's value for the first metric in
+                    `metric_list`.
+                - ...
+                - [metric n]: The client's value for the nth (final)
+                    metric in `metric_list`.
+                - [sanity check 1]: The client's value for the first
+                    sanity check metric.
+                - ...
+                - [sanity check n]: The client's value for the last
+                    sanity check metric.
+
+            This format - the schema plus there being one row per
+            enrolled client, regardless of whether the client has data
+            in `data_source` - was agreed upon by the DS team, and is the
+            standard format for queried experimental data.
+        """
+        for col in ['client_id', 'submission_date_s3']:
+            if col not in data_source.columns:
+                raise ValueError("Column '{}' missing from 'data_source'".format(col))
+
+        req_dates_of_data = conv_window_start_days + conv_window_length_days
+
+        if self.num_dates_enrollment is None:
+            self._print_enrollment_window(last_date_full_data, req_dates_of_data)
+
+        enrollments = self.filter_enrollments_for_conv_window(
+            enrollments, last_date_full_data, req_dates_of_data
+        )
+
+        data_source = self.filter_data_source_for_conv_window(
+            data_source, last_date_full_data, conv_window_start_days,
+            conv_window_length_days
+        )
+
+        join_on = [
+            # TODO perf: would it be faster if we enforce a join on sample_id?
+            enrollments.client_id == data_source.client_id,
+
+            # TODO accuracy: once we can rely on
+            #   `data_source.experiments[self.experiment_slug]`
+            # existing even after unenrollment, we could start joining on
+            # branch to reduce problems associated with split client_ids:
+            # enrollments.branch == data_source.experiments[self.experiment_slug]
+
+            # Do a quick pass aiming to efficiently filter out lots of rows:
+            enrollments.enrollment_date <= data_source.submission_date_s3,
+
+            # Now do a more thorough pass filtering out irrelevant data:
+            # TODO perf: what is a more efficient way to do this?
+            (
+                (
+                    F.unix_timestamp(data_source.submission_date_s3, 'yyyyMMdd')
+                    - F.unix_timestamp(enrollments.enrollment_date, 'yyyyMMdd')
+                ) / (24 * 60 * 60)
+            ).between(
+                conv_window_start_days,
+                conv_window_start_days + conv_window_length_days - 1
```

Comment:
do you exclude the last day of the conversion window?

Reply:
`between` is inclusive, so this should obtain data from `conv_window_length_days` consecutive submission dates, unless I stuffed it up?

I frequently fall into this trap or at the very least second-guess myself here, so I'm looking forward to hiding these choices behind an interface that accepts an easily interpretable quantity (`conv_window_length_days`), and has automated tests with examples of how it should behave.

Reply:
I think that if it is inclusive then it should be `conv_window_start_days + conv_window_length_days` otherwise it is clipping unless I am missing something.

Reply:
If you set `conv_window_length_days = 1`, then how many dates of data do you expect to be returned? Would you expect data collected from one date or two? I would expect one only, and that's what I think this code does. Note that we're comparing dates, not real timestamps.

There are some examples that test this join [here](https://github.com/felixlawrence/mozanalysis/blob/revamp-and-add-query-code/tests/test_experiment.py#L308) though I'll admit that this is the one of the most convoluted parts of the test suite because it's testing something in the middle of a function

Reply:
ah, sorry, I get it now. I was thinking in python indexing terms instead of the dates

Reply:
This is precisely the kind of thing that I commonly stuff up, so I'm grateful for the extra eyes :)